### ISSUE TITLE
AMP's protocol switch feature can deliver an empty string to a switched-to protocol's dataReceived

### ISSUE THREAD

**Issue by twisted-trac** [twisted/twisted#4405]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)| @exarkun reported|
|-|-|
|Trac ID|trac#4405|
|Type|defect|
|Created|2010-04-17 17:14:42Z|
|Branch|https://github.com/twisted/twisted/tree/empty-datareceived-4405|

That the string passed to `Protocol.dataReceived` won't be empty is sort of a guarantee made by the transport/protocol interface.  This isn't implemented by AMP's protocol switch feature, though.  After switching to a new protocol, AMP will unconditionally deliver a string of bytes to the new protocol's `dataReceived` method, even if there are no bytes left in the input buffer - resulting in `dataReceived` being called with an empty string sometimes.

It should check first and skip the `dataReceived` call if there's no data.

An example of some real world code that this breaks is Q2Q.

<details><summary>Searchable metadata</summary>

```
trac-id__4405 4405
type__defect defect
reporter__exarkun exarkun
priority__normal normal
milestone__ 
branch__branches_empty_datareceived_4405 branches/empty-datareceived-4405
branch_author__exarkun exarkun
status__closed closed
resolution__fixed fixed
component__core core
keywords__amp amp
time__1271524482000000 1271524482000000
changetime__1273968106000000 1273968106000000
version__None None
owner__ 
cc__rwall
```
</details>

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_1"></a>|@exarkun commented|
|-|-|

(In [28830]) Branching to 'empty-datareceived-4405'

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)|@exarkun removed owner|
|-|-|

**Comment by twisted-trac** [comment]
|[<img alt="wallrj's avatar" src="https://avatars.githubusercontent.com/u/978965?s=50" width="50" height="50">](https://github.com/wallrj)<a name="note_3"></a>|@wallrj set owner to @exarkun|
|-|-|

Looks good.
 * New test and newsfile make sense
 * confirmed that new test fails against trunk and passes in the branch
 * confirmed that all amp tests still pass in the branch

Please merge.

**Comment by twisted-trac** [comment]
|[<img alt="exarkun's avatar" src="https://avatars.githubusercontent.com/u/254565?s=50" width="50" height="50">](https://github.com/exarkun)<a name="note_4"></a>|@exarkun set status to `closed`|
|-|-|

(In [28948]) Merge empty-datareceived-4405

Author: exarkun
Reviewer: rwall
Fixes: [#4405](https://github.com/twisted/twisted/issues/4405)


Fix a case where `twisted.protocols.amp.BinaryBoxProtocol` would sometimes
delivery an empty string to another protocol's `dataReceived` method after
"switching" to that protocol.

**Comment by twisted-trac** [comment]
|<img alt="Automation's avatar" src="https://avatars.githubusercontent.com/u/0?s=50" width="50" height="50">|Automation removed owner|
|-|-|
